MelodyGenerator.js

```diff
@@ -0,0 +1,145 @@
+const durations = require('./durations.js');
+const MelodyQueue = require('./MelodyQueue.js');
+const waveGenerator = require('./WaveGenerator.js');
+const note = require('./note.js');
+
+//Global variables that are accessible to all
+let melody = new MelodyQueue();
+let macros = new Map(); // In case a macro needs to be defined
+const regex1 = /repeat (\d) times (.+)/; // Regex for repeat commands
+const regex2 = /((note\(|pause\()((.{2,3})(;(\w\w))*)(,\s?([A-Z]))?\))/; //Regex for notes, pauses
+const regex3 = /define (\w+) is (.+)/;  //Regex for defining macros
+const regex4 = /(play )?(.+)/; // Regex for playing non-repeating macros
+
+class MelodyGenerator
+{
+
+	parseMacro(lineString) {
+		// define mel01 is pause(HT) | note(E5, E)
+		const parsedMacro = regex3.exec(lineString); 
+		const macroName = parsedMacro[1];
+		const line = parsedMacro[2].split("|");
+		let macroNotes = [];
+
+		for(let i=0; i < line.length; i++) {
+				macroNotes = [...macroNotes, ...this.parseSound(line[i])];
```

Comment:
Of course I tried to push rather than concatenating with spreads, but the problem was I was pushing an array of arrays into an array, which made the interpretation of that array difficult to loop through (I guess I could have tried recursion but I didn't want to go there...) so I added the spread to empty the contents of the array into the existing array. The end results was something like this [[E, G4], [Q, C4], [HT]] rather than [[E, G4], [[Q, C4], [HT]]]. See what I mean?

forEach? Yes, this is an option but weirdly I don't have a lot of experience using it in Javascipt, it's more abundant in Ruby, for example. I'll make this modification.

Reply:
`this.parseSound` returns array? Oh, my bad. So you can use concat:
`macroNotes = macroNotes.concat(this.parseSound(line[i])`